Disable using driver after driver is disposed

zhenlineo opened the pull request:
Made the driver to be thread safe for concurrent access on driver.dispose and driver.session

Commit:
Disable using driver after driver is disposed
Made the driver to be thread safe

Commit:
Avoid NPE because of setting unmanaged resources to be null
Make sure no new conn resource could be open after pool is disposed (ClusterConnectionPool.Add)
Add more logging near the code for updating routing tables

Commit:
Removed the lock on driver.dispose to protect against concurrent call of driver.dispose
Because it is okay to call driver.close multiple times
Fixed the failing tests

lutovich commented on Neo4j.Driver/Neo4j.Driver/Internal/Routing/ClusterDiscoveryManager.cs:
```diff
@@ -41,30 +41,41 @@ public ClusterDiscoveryManager(IConnection connection, ILogger logger)
-                var record = result.Single();
-                // TODO require an IT to make sure List or IList, Dictionary or IDictionary
-                foreach (var servers in record["servers"].As<List<Dictionary<string,object>>>())
+                var session = new Session(_conn, _logger);
```
There should probably be a `finally` block where session is closed.

lutovich commented on Neo4j.Driver/Neo4j.Driver/Internal/Routing/ConcurrentRoundRobinSet.cs:
```diff
@@ -107,6 +107,21 @@ public bool TryNext(out T value)
+        {
+            lock (_syncLock)
+            {
+                while (!_queue.IsEmpty)
```
This could alternatively be written as:

```
T ignored;
while (_queue.TryDequeue(out ignored)) {}
```

lutovich commented on Neo4j.Driver/Neo4j.Driver/Internal/Routing/RoundRobinLoadBalancer.cs:
```diff
@@ -166,19 +166,22 @@ internal RoundRobinRoutingTable UpdateRoutingTable(Func<IPooledConnection, Round
-                    // The result is invalid probably due to partition.
-                    _routingTable.Remove(uri);
+                    _logger?.Info($"Failed to update routing table with server uri={uri} due to error {e.Message}");
+                    if (e is SessionExpiredException)
```
Can this branch be removed? It seems like we are only interested in `InvalidDiscoveryException` here.

lutovich commented on Neo4j.Driver/Neo4j.Driver/Internal/Routing/RoundRobinRoutingTable.cs:
```diff
@@ -54,7 +54,7 @@ public RoundRobinRoutingTable(Uri seed, Stopwatch stopwatch, long expireAfterSec
         public bool IsStale()
         {
-            return _routers.Count <= MinRouterCount || _readers.IsEmpty || _writers.IsEmpty
+            return _routers.Count < MinRouterCount || _readers.IsEmpty || _writers.IsEmpty
```
Could we please have a test for this change?

lutovich reviewed (changes requested):
@zhenlineo review completed. Made couple comments.

Commit:
Fix after code review
Added back using while still keep try outside to catch any error arise inside using
Updated the clear method
Added more tests

zhenlineo commented:
Merging the PR after it turns to green
